src/Agent.Worker/Variables.cs

```diff
@@ -221,140 +228,155 @@ public bool TryGetValue(string name, out string val)
             return false;
         }
 
-        private void RecursivelyExpand(out List<string> warnings)
+        public void RecalculateExpanded(out List<string> warnings)
```

Comment:
Very little actually changed within this method.
1) I moved the code within a critical section (using `_setLock` - the same mutex used by the `Set()` method).
2) Rather than copying `_store` into `original`, I use `_nonexpanded` as the source dictionary and new-up a local `expanded` dictionary. At the end of the method, `expanded` is stored to `_expanded`.
3) When the fully expanded value for a variable is determined, I do not call `Set()` anymore. Instead I store it to the local expanded dictionary and register the secret. `Set()` modifies the nonexpanded dictionary as well, and we don't want to do that.